model/map/src/main/java/org/keycloak/models/map/storage/chm/SingleUseObjectModelCriteriaBuilder.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.chm;
+
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.common.StringKeyConverter;
+import org.keycloak.storage.SearchableModelField;
+
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
+ */
+public class SingleUseObjectModelCriteriaBuilder<K, V extends AbstractEntity, M> extends MapModelCriteriaBuilder<K, V, M> {
+
+    private String userId;
+
+    private String actionId;
+
+    private String actionVerificationNonce;
+
+    StringKeyConverter<K> keyConverter;
+
+    Map<SearchableModelField<? super M>, UpdatePredicatesFunc<K, V, M>> fieldPredicates;
+
+    public SingleUseObjectModelCriteriaBuilder(StringKeyConverter<K> keyConverter,
+                                               Map<SearchableModelField<? super M>, UpdatePredicatesFunc<K, V, M>> fieldPredicates) {
+        this(keyConverter, fieldPredicates, ALWAYS_TRUE, ALWAYS_TRUE);
+    }
+
+    protected SingleUseObjectModelCriteriaBuilder(StringKeyConverter<K> keyConverter, Map<SearchableModelField<? super M>, UpdatePredicatesFunc<K, V, M>> fieldPredicates, Predicate<? super K> indexReadFilter, Predicate<? super V> sequentialReadFilter) {
+        super(keyConverter, fieldPredicates, indexReadFilter, sequentialReadFilter);
+        this.keyConverter = keyConverter;
+        this.fieldPredicates = fieldPredicates;
+    }
+
+    @Override
+    public MapModelCriteriaBuilder<K, V, M> compare(SearchableModelField<? super M> modelField, Operator op, Object... values) {
+        if (modelField == org.keycloak.models.ActionTokenValueModel.SearchableFields.USER_ID) {
+            userId = values[0].toString();
+        } else if (modelField == org.keycloak.models.ActionTokenValueModel.SearchableFields.ACTION_ID) {
+            actionId = values[0].toString();
+        } else if (modelField == org.keycloak.models.ActionTokenValueModel.SearchableFields.ACTION_VERIFICATION_NONCE) {
+            actionVerificationNonce = values[0].toString();
+        }
+        return super.compare(modelField, op, values);
```

Comment:
This is a little bit confusing. Do we want to support other queries than `USER_ID && ACTION_ID &&  ACTION_VERIFICATION_NONCE`? It seems `SingleUseObjectModelCriteriaBuilder` wants to allow that by including `fieldPredicates`, `indexReadFilter`, `sequentialReadFilter` in the constructor, however, I don't see how this could work as using parent compare method basically removes all fields from this object.

I would say we should start with supporting only these three parameters and refactor in the future if we need more attributes. We should probably stop extending `MapModelCriteriaBuilder` and implement only `MapModelCriteriaBuilder` interface, this way we would get rid of `fieldPredicates`, `indexReadFilter` and `sequentialReadFilter` completely.